File: Math4610_local/bin/data/doc/inc/lib/Homework_2/Homework_2.c
#include <stdio.h>
#include <math.h>
#include <stdlib.h>

void machineEps32();
void machineEps64();
double l_2_norm(double vector[], int size);
double l_1_norm(double vector[], int size);
double infinity_norm(double vector[], int size);
double l_2_distance(double u[], double v[], int size);
double l_1_distance(double u[], double v[], int size);
double infinity_norm_distance(double u[], double v[], int size);

int main() {
    // machineEps32();
    // machineEps64();
    double u[] = {2.0, 2.0};
    double v[] = {3.0, -1000.0};
    // l_2_norm(u, sizeof(u)/sizeof(double));
    // l_1_norm(u, sizeof(u)/sizeof(double));
    // infinity_norm(v, sizeof(v)/sizeof(double));
    // l_2_distance(u, v, sizeof(u)/sizeof(double));
    // l_1_distance(u, v, sizeof(u)/sizeof(double));
    // infinity_norm_distance(u, v, sizeof(u)/sizeof(double));
    return 0;
}

void machineEps32() {
    float previous_eps = 0.0f;
    float eps = 1.0f;
    while ((1.0f + eps) != 1.0f){
        previous_eps = eps;
        eps /= 2.0f;
    }
    printf("%e\n", previous_eps);
}

void machineEps64() {
    double previous_eps = 0.0;
    double eps = 1.0;
    while ((1.0 + eps) != 1.0){
        previous_eps = eps;
        eps /= 2.0;
    }
    printf("%e\n", previous_eps);
}

double l_2_norm(double *vector, int size) {
    double length = 0;
    for(int i = 0; i < size; i++) {
        length += pow(vector[i], 2);
    }
    printf("%lf\n", sqrt(length));
    return sqrt(length); 
}

double l_1_norm(double *vector, int size) {
    double length = 0;
    for(int i = 0; i < size; i++) {
        length += fabs(vector[i]);
    }
    printf("%lf\n", length);
    return length;
}

double infinity_norm(double *vector, int size) {
    double max = vector[0];
    for(int i = 0; i < size; i++) {
        double abs_i = fabs(vector[i]);
        if (max < abs_i) {
            max = abs_i;
        }
    }
    return max;
}

double l_2_distance(double *u, double *v, int size) {
    double distance = 0;
    for(int i = 0; i < size; i++) {
        distance += pow((u[i] - v[i]), 2);
    }
    printf("%lf\n", sqrt(distance));
    return sqrt(distance);
}

double l_1_distance(double *u, double *v, int size) {
    double distance = 0;
    for(int i = 0; i < size; i++) {
        distance += fabs(u[i] - v[i]);
    }
    printf("%lf", distance);
    return distance;
}

double infinity_norm_distance(double u[], double v[], int size) {
    double compute = infinity_norm(u, size) - infinity_norm(v, size);
    return fabs(compute);
}

File: Math4610_local/bin/data/doc/inc/lib/Homework_2/Homework_2.py
import math as m
import numpy as np

"""Homework 1"""

#### Question 1
# a. Machine precision value in single precision. float
# Python does not have a 32-bit float
print()
print("Question 1")
print("a. float precision: ")
print(np.finfo(np.float32).eps)

# b. Machine precision value in double precision. double
# Note that Python's float is actually a double (64-bit)
previous_eps = 0.0
eps = 1.0
while (1.0 + eps != 1.0):
    previous_eps = eps
    eps /= 2.0
print("b. double precision: ")
print(previous_eps)
    

### Question 2
#a. return length of a vector
def l_2_nom(vector):
    length = 0
    for i in vector:
        length += i**2
    return m.sqrt(length)

#b. return l-norm
def l_1_norm(vector):
    length = 0
    for i in vector:
        length += abs(i)
    return length

#c. return infinity norm
def infinity_norm(vector):
    max = vector[0]
    for i in vector:
        abs_i = abs(i)
        if max < abs_i:
            max = abs_i
    return max


### Question 3
#a. distance between 2 vectors in l2
def l2_distance(u, v):
    distance = 0
    for i, j in zip(u, v):
        distance += (i - j)**2
        
    return distance

#b. distance between 2 vectors in l1
def l1_distance(u, v):
    distance = 0
    for i, j in zip(u, v):
        distance += abs(i - j)  
    return distance

#c. distance between 2 infinity vectors
def infinity_distance(u, v):
    return abs(infinity_norm(u) - infinity_norm(v))


def report():    
    # print()
    # print("Question 2")
    # print("a. l2 norm: ")
    # print(l_2_nom([1,2,3]))
    # print(l_2_nom([2,-2,2,-2]))
    # print(l_2_nom([0,5,-11,15,33]))
    # print("b. l1 norm: ")
    # print(l_1_norm([1,2,3]))
    # print(l_1_norm([2,-2,2,-2]))
    # print(l_1_norm([0,5,-11,15,33]))
    # print("c. infinity norm: ")
    # print(infinity_norm([3]))
    # print(infinity_norm([1,100,11]))
    # print(infinity_norm([30,-1001,200]))
    # print()
    # print("Question 3")
    # print("a. l2 distance: ")
    # print(l2_distance([3], [4]))
    print(l2_distance([2,2], [-1000,3]))
    print(l1_distance([2,2], [-1000,3]))
    # print(l2_distance([1,5,3], [3,2,1]))
    # print("b. l1 distance: ")
    # print(l1_distance([1], [3]))
    # print(l1_distance([1,2], [3,4]))
    # print(l1_distance([1,5,3], [3,2,1]))
    # print("c. infinity distance: ")
    # print(infinity_distance([1], [3]))
    # print(infinity_distance([1,7], [3,1]))
    # print(infinity_distance([1,2,3], [100, 1000, 10000]))
    
    
### MAIN ###
# Calls the report function to show computations
def main():
    report()

if __name__ == "__main__":
    main()
        
    
    